### PR TITLE
App registration without tests

### PR DESCRIPTION
**Pull request by prakharjoshi**

### CODE REVIEW

File: src/experimental/safe_html_transform/configure.zcml [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
     <include package=".browser" />
 
+    <include package="plone.app.registry" />
```
Comment (tisto) [inline src/experimental/safe_html_transform/configure.zcml]: This line shouldn't be necessary.

File: src/experimental/safe_html_transform/interfaces.py [hunk 2]
```diff
@@ -2,6 +2,34 @@
...
+ 
+_ = MessageFactory('experimental.safe_html_transform')
+ 
+ 
+class IExperimenatalSafe_Html_TransformSettings(Interface):
+"""Global safe_html_transform settings. This describes records stored in the
+	configuration registry and obtainable via plone.registry.
+"""
+ 
+    Experimentalsafe_html_transform_key = schema.TextLine(title=_(u"Experimenatalsafe_html_transform (Wordpress) Key"),
```
Comment (tisto) [inline src/experimental/safe_html_transform/interfaces.py]: There is no need to use an "Experimentalsafe_html" prefix for everything. Also you should use the schema from the CMFPlone pull request in the Google docs document.

Comment (prakharjoshi) [inline src/experimental/safe_html_transform/interfaces.py]: "Also you should use the schema from the CMFPlone pull request in the
Google docs document." Didn't get that part ?

Prakhar Joshi
DA-IICT,Gandhinagar

Comment (tisto) [inline src/experimental/safe_html_transform/interfaces.py]: You should basically just copy the code from this pull request into our new package: https://github.com/plone/Products.CMFPlone/commit/ac98b0ea4ae27f4b74f4d3c5cf100b54f8acbfb7

Comment (prakharjoshi) [inline src/experimental/safe_html_transform/interfaces.py]: here its seems like the contropanel is a folder containing different files.
Should we go like that ?

Prakhar Joshi
DA-IICT,Gandhinagar

Comment (tisto) [inline src/experimental/safe_html_transform/interfaces.py]: It doesn't matter in Plone if those files are in a different folder. You can keep your current structure. You just have to replace the Class. Though, you need to understand what you are doing. If you are just copy/pasting code it won't work. See https://pypi.python.org/pypi/plone.app.registry for details about the registry control panels.

Comment (prakharjoshi) [inline src/experimental/safe_html_transform/interfaces.py]: We should also create a module for controlpanel and then we can work
accordingly ?

Prakhar Joshi
DA-IICT,Gandhinagar

Comment (prakharjoshi) [inline src/experimental/safe_html_transform/interfaces.py]: Yeah I know what is going on there just want to conform that either we go
with current structure or we should change according to the commit one. But
yeah got my answer. Will correct that :)

Prakhar Joshi
DA-IICT,Gandhinagar

Comment (prakharjoshi) [inline src/experimental/safe_html_transform/interfaces.py]: hi,
    I just created a pull request
https://github.com/collective/experimental.safe_html_transform/pull/1 .
Just adding the filter in the current project without breaking the basic
structure of the app.

Cheers!!

Prakhar Joshi
DA-IICT,Gandhinagar

**Comment by tisto** [comment]
@prakharjoshi Your commit fails both code-analysis and the tests. Please run "bin/test --all" and "bin/code-analysis" and make sure they both pass before you push any commit.

**Comment by gforcada** [comment]
@prakharjoshi would you mind having a look at the output of [travis](https://travis-ci.org/collective/experimental.safe_html_transform/builds/50215198)? There are quite a few code analysis related errors that prevent tests to run

**Comment by prakharjoshi** [comment]
hi,
    I corrected the code and seems like all the test cases run successfully
and also the code analysis ends with '0' . I pushed the code also
https://github.com/collective/experimental.safe_html_transform/pull/1 . But
it seems like there is an error in travis
https://travis-ci.org/collective/experimental.safe_html_transform/builds/50328240
. Don't know why ?

Cheers!!

Prakhar Joshi
DA-IICT,Gandhinagar

**Comment by gforcada** [comment]
This error is due to a newer setuptools version (12.1) which seems to break something... let's pin it at 8.3...

**Comment by gforcada** [comment]
Adding a [version] section on travis.cfg stating "setuptools = 8.3" is most probably enough :)

**Comment by prakharjoshi** [comment]
@Gil Forcada . I did that correction here
https://github.com/collective/experimental.safe_html_transform/pull/1 But
it is showing the same error
https://travis-ci.org/collective/experimental.safe_html_transform/builds/50336959
.

Prakhar Joshi
DA-IICT,Gandhinagar

**Comment by gforcada** [comment]
Sorry there's another thing that need to be fixed: on .travis.yml you need to fix the [bootstrap line](https://github.com/collective/experimental.safe_html_transform/blob/master/.travis.yml#L6) to also add the argument `--setuptools-version=8.3`

**Comment by prakharjoshi** [comment]
@gforcada  thanks, it works!! Finally a correct tests passed and travis approved commit.

**Comment by bogdangi** [comment]
@prakharjoshi , I can't see a link to the successful travis build, could you please send me the link?

**Comment by prakharjoshi** [comment]
@bogdangi  Sorry for that, here is the link https://github.com/prakharjoshi/experimental.safe_html_transform/commit/0106f87a2b636eb3af19cf48992c8fee945e2bbe

**Comment by bogdangi** [comment]
@prakharjoshi , that link to the commit not to **the successful travis build**

**Comment by prakharjoshi** [comment]
@bogdangi  Oh.. here it is https://github.com/prakharjoshi/experimental.safe_html_transform/tree/0106f87a2b636eb3af19cf48992c8fee945e2bbe

**Comment by bogdangi** [comment]
@prakharjoshi , I believe that the link to successful travis build should start from `https://travis-ci.org/collective/experimental.safe_html_transform/builds/`

**Comment by prakharjoshi** [comment]
@bogdangi  I can't find the link for that commit as previously if there was a travis failure the there was a 'x' mark in red with the commit number but in my last commit there is no such mark with the commit number. So I though it passed the travis.

**Comment by gforcada** [comment]
@prakharjoshi for some reason your last commit did not trigger a new build on travis, would you mind doing a random change on a file and pushing it again?

You can see the x mark if you click on the "commits" tab above these messages.

**Comment by prakharjoshi** [comment]
@gforcada  I pushed a new dummy commit but still it is not traced by travis :( .  Is that due to change in .travis.yml file ?

**Comment by gforcada** [comment]
@prakharjoshi maybe as you were changing .travis.yml travis didn't like it, try changing another file.

btw, you should squash together the commits where you added LOTS of files and on the later one that you removed it. If you do a new clone it takes ages because of those many files... See a tutorial on how to do that: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase-i/

An an example on how it should end up looking (note that I also squashed together other commits): https://github.com/gforcada/experimental.safe_html_transform/commits/master

**Comment by tisto** [comment]
@gforcada what about adding docs for squashing commits to the buildout.coredev jobs? or maybe an entire section about git best practices?

**Comment by gforcada** [comment]
@tisto yes, I've been thinking on that, I will try to compile some links and best practices.

File: src/experimental/safe_html_transform/browser/configure.zcml [hunk 3]
```diff
@@ -17,4 +17,14 @@
         directory="static"
         />
 
+    <adapter factory=".controlpanel.FilterControlPanelAdapter" />
+
+    <!-- Filter Control Panel -->
+    <browser:page
+      name="controlpanel"
```
Comment (tisto) [inline src/experimental/safe_html_transform/browser/configure.zcml]: @prakharjoshi You registered the filter control panel as "controlpanel" and the configlet with "safe_html_transform-settings". I do not think that works at all. Have you tested this code? We should either override the filter-controlpanel view or add an additional view.

**Comment by prakharjoshi** [comment]
@tisto  All the test cases are passed. have a look http://pastebin.com/uhV6J8eU

**Comment by gforcada** [comment]
@prakharjoshi no need to paste your tests, that's what travis-ci.org is for.

You can see that the green ticks are back on your commits, so travis is running again.

And indeed, all your tests pass :)

**Comment by prakharjoshi** [comment]
The "for" used under the "<browser:page" is used for what ?

**Comment by gforcada** [comment]
That `for` is used to register the given view that you are declaring to be shown on objects that have that interface, so in your registration on any object that has `.controlpanel.IPloneSiteRoot` interface you can append a `/@@controlpanel` and you view will be rendered.

So as you are subclassing `ISiteRoot` and `INavigationRoot` this means that directly on the plone instance home page you can append `/@@controlpanel`.

Does that make it clear?

**Comment by prakharjoshi** [comment]
@gforcada Yeah I got that. One more thing what is the use of layer under "<browser:page"  ?

**Comment by gforcada** [comment]
@prakharjoshi the layer is used as an override mechanism, if there are two browser views registered for the same interface (on "for") and also with the same name (also "name" on browser:page) zope takes a look at the layer and sorts it by specificity, your own layer will be always above a core plone layer so to say.

File: src/experimental/safe_html_transform/browser/configure.zcml [hunk 4]
```diff
@@ -17,4 +17,15 @@
         directory="static"
         />
 
+    <adapter factory=".controlpanel.FilterControlPanelAdapter" />
```
Comment (tisto) [inline src/experimental/safe_html_transform/browser/configure.zcml]: @prakharjoshi I think you can remove the FilterControlPanelAdapter. This is required to keep backwards compatibility in Plone core only. We don't need it in an add-on product. Remove that line and the FilterControlPanelAdapter class in controlpanel.py as well.

File: src/experimental/safe_html_transform/profiles/default/controlpanel.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
```
Comment (tisto) [inline src/experimental/safe_html_transform/profiles/default/controlpanel.xml]: I don't think this is necessary here. You are registering an non-existent safe_html_transform control panel. we are re-using the @@filter-controlpanel.

File: buildout.cfg [hunk 6]
```diff
@@ -1,2 +1,2 @@
 [buildout]
-extends = buildout.d/development.cfg
+extends = buildout.d/development.cfg
```
Comment (lentinj) [inline buildout.cfg]: buildout.cfg lost a newline at the end of the file, can we have it back?

File: src/experimental/safe_html_transform/browser/controlpanel.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
+from Products.CMFPlone import PloneMessageFactory as _
+from plone.app.registry.browser import controlpanel
+# from Products.CMFPlone.interfaces import IFilterSchema
+from Products.CMFCore.interfaces import ISiteRoot
+from plone.app.layout.navigation.interfaces import INavigationRoot
+# from plone.registry.interfaces import IRegistry
+from zope.interface import Interface
+# from zope.component import adapts
+from zope import schema
+# from zope.component import getUtility
```
Comment (lentinj) [inline src/experimental/safe_html_transform/browser/controlpanel.py]: If you don't need them, delete them, don't comment them out. You can always find them again with the git history. Nothing is lost forever :)

File: buildout.cfg (hunk 6 above)
Comment (prakharjoshi) [inline buildout.cfg]: @lentinj  Yeah Will make it correct in new push.

File: src/experimental/safe_html_transform/browser/controlpanel.py (hunk 7 above)
Comment (prakharjoshi) [inline src/experimental/safe_html_transform/browser/controlpanel.py]: Okay. Will clean this thing and will take care of that in future :)

File: src/experimental/safe_html_transform/interfaces.py [hunk 8]
```diff
@@ -2,6 +2,9 @@
 """Module where all interfaces, events and exceptions live."""
 
 from zope.publisher.interfaces.browser import IDefaultBrowserLayer
+# from z3c.form import interfaces
+from zope.i18nmessageid import MessageFactory
+_ = MessageFactory('experimental.safe_html_transform')
```
Comment (lentinj) [inline src/experimental/safe_html_transform/interfaces.py:6]: Generally Plone add-ons create one MessageFactory and import that everywhere. For example:-

https://github.com/collective/collective.alias/blob/master/src/collective/alias/interfaces.py#L8

The bobtemplate made this for you, so you just have to do `from experimental.safe_html_transform import MessageFactory as _` here.

File: src/experimental/safe_html_transform/profiles/default/metadata.xml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <metadata>
   <version>1000</version>
-  <dependencies>
+  <dependencies>profile-plone.app.registry:default
```
Comment (lentinj) [inline src/experimental/safe_html_transform/profiles/default/metadata.xml]: Dependencies go inside a `<dependency>` node. Have a look at:-

https://github.com/collective/collective.alias/blob/master/src/collective/alias/profiles/default/metadata.xml#L4

File: src/experimental/safe_html_transform/configure.zcml [hunk 10]
```diff
@@ -17,6 +17,15 @@
...
 
+    <genericsetup:registerProfile
+      name="uninstall"
+      title="experimental.safe_html_transform uninstall"
+      directory="profiles/uninstall"
+      description="Uninstalls the experimental.safe_html_transform package"
+      provides="Products.GenericSetup.interfaces.EXTENSION"
+      />
+
+
```
Comment (lentinj) [inline src/experimental/safe_html_transform/configure.zcml:28]: This is good stuff---plenty of people will be very happy to see you thinking of an uninstall profile already, but there isn't a `src/experimental/safe_html_transform/profiles/uninstall` directory just yet. Don't register it until you've got something useful there.

**Comment by lentinj** [comment]
There's lots of good bits in here, but they're added before their time. Obviously it's tricky while you're getting an add-on started, but generally a pull request should represent a complete set of changes that are ready to go.

It's not a huge deal at this point, but I'd be tempted to start a new branch (let me know if you want git specifics) that picks out the changes that are definitely going to get used. For example, the setuptools version 8.3 changes. Then changes to add the registry settings. Then the control panel overriding on top of that.

I bring it up as it would also mean that we don't have the eggs added / removed in the git history. Which is currently making the history for this repository a heafty download :)

File: src/experimental/safe_html_transform/profiles/default/metadata.xml (hunk 9 above)
Comment (prakharjoshi) [inline src/experimental/safe_html_transform/profiles/default/metadata.xml]: <dependency> caused travis failure, I guess <dependencies> will work here ?

File: src/experimental/safe_html_transform/configure.zcml (hunk 10 above)
Comment (prakharjoshi) [inline src/experimental/safe_html_transform/configure.zcml:28]: @tisto @lentinj  Can you provide a sample example of add-on where uninstall profile is used ?

**Comment by prakharjoshi** [comment]
@lentinj  Okay lets start with a new branch and yes sure just provide me the git specifics (will be helpful). Now we will take care of new branch ane try to keep that clean :)

**Comment by prakharjoshi** [comment]
@lentinj @tisto  Have a look at the uninstall profile, And I am not able to detect why travis is not working :( .

**Comment by tisto** [comment]
@prakharjoshi travis is working and the build passes: https://travis-ci.org/collective/experimental.safe_html_transform/builds/62255709

**Comment by prakharjoshi** [comment]
@tisto then why shouldn't a travis link appearing against my last commit ?

**Comment by prakharjoshi** [comment]
@tisto @lentinj  Here is the integrated version of transform and it is failing for 3 test cases as present https://travis-ci.org/collective/experimental.safe_html_transform/builds/65390035

**Comment by prakharjoshi** [comment]
@tisto  Have a look at the travis result, test cases are passing here in local, I guess here is some version problem, right ?

**Comment by tisto** [comment]
Yeah. It seems a package we use is fetching the latest Plone 5 release.

**Comment by prakharjoshi** [comment]
Okay so we have to pin up the version to plone 4.3.x or we should change settings so that it works fo plone 5 new releases ?

**Comment by tisto** [comment]
We have to pin versions to Plone 4.3.x. See: http://maurits.vanrees.org/weblog/archive/2008/01/easily-creating-repeatable-buildouts

Later we can create a second travis job that tests against Plone 5.0.x

**Comment by prakharjoshi** [comment]
Sure. Will modify the buildout.cfg :)

**Comment by lentinj** [comment]
I'm assuming this pull request is dead now.

**Pull request closed without merging**